lib/mongoid/contextual/mongo.rb

```diff
@@ -1038,6 +1043,24 @@ def process_raw_docs(raw_docs, limit)
         limit ? docs : docs.first
       end
 
+      # Queries whether the current context is valid for use with
+      # the #count_documents? predicate. A context is valid if it
+      # does not include a `$where` operator.
+      #
+      # @return [ true | false ] whether or not the current context
+      #   excludes a `$where` operator.
+      def valid_for_count_documents?(hash = view.filter)
+        # Note that `view.filter` is a BSON::Document, and all keys in a
+        # BSON::Document are strings; we don't need to worry about symbol
+        # representations of `$where`.
+        hash.keys.each do |key|
+          return false if key == '$where'
+          return false if hash[key].is_a?(Hash) && !valid_for_count_documents(hash[key])
```

Comment:
I ran a quick benchmark on recursively traversing a `Hash` versus `hash.to_s.match?`, and it was quite a bit faster to traverse the hash, even in the pathological cases (deeply nested with no matching key). It wasn't that much harder to do the recursive traversal, so I opted for that.

/cc @alcaeus @alexbevi